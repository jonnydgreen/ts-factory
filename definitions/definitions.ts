import { BlockInput } from './block/block.type.ts';
import { FunctionDeclarationInput } from './function-declaration/function-declaration.type.ts';
import { IdentifierInput } from './identifier/identifier.type.ts';
import { KeywordTypeNodeInput } from './keyword-type-node/keyword-type-node.type.ts';
import { ModifierInput } from './modifier/modifier.type.ts';
import { SourceFileInput } from './source-file/source-file.type.ts';

// Definition
export type Definition =
  | SourceFileInput
  | StatementInput
  | ExpressionInput
  | TypeNodeInput
  | ModifierInput;

// Member Name
export type MemberNameInput = IdentifierInput;

// Statement Inputs
export type StatementInput = FunctionDeclarationInput | BlockInput;

// Expression Inputs
export type ExpressionInput = IdentifierInput;

// Type Node Inputs
export type TypeNodeInput = KeywordTypeNodeInput;

// Input
export type Input = SourceFileInput;
