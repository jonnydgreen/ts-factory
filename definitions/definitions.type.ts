import type { ts } from '../deps.ts';
import type { Instructions } from '../instructions/instructions.type.ts';
import type {
  ExpressionInput,
  MemberNameInput,
  PropertyNameInput,
  StatementInput,
  TypeElementInput,
  TypeNodeInput,
} from './definitions.ts';
import { ModifierInput } from './modifier/modifier.type.ts';
import { TokenInput } from './token/token.type.ts';
import type { LeadingTriviaInput } from './trivia/trivia.type.ts';

export type MergeIntersections<I> = { [K in keyof I]: I[K] };

export type NonFnKeys<T> = {
  // deno-lint-ignore no-explicit-any
  [K in keyof T]: T[K] extends (...args: any[]) => any ? never : K;
}[keyof T];

// deno-fmt-ignore
export type MapTSTypes<T> =
  T extends ts.SyntaxKind ? T :
  T extends ts.NodeArray<infer TNode> ? MapTSTypes<TNode>[] :
  T extends ts.__String ? string :
  T extends string ? string :
  T extends ts.Statement ? StatementInput :
  T extends string | ts.MemberName ? string | MemberNameInput :
  T extends string | ts.PropertyName ? string | PropertyNameInput :
  T extends ts.Modifier ? ModifierInput :
  T extends ts.TypeNode ? TypeNodeInput :
  T extends ts.TypeElement ? TypeElementInput :
  T extends ts.Expression ? ExpressionInput :
  T extends ts.SyntaxKind ? ExpressionInput :
  T extends ts.Token<ts.SyntaxKind> ? TokenInput :
  unknown;

export type CreateInput<
  TNode extends ts.Node,
  TNodeKeys extends NonFnKeys<TNode>,
  TPartialNodeKeys extends NonFnKeys<TNode> = TNodeKeys,
> = MergeIntersections<
  & {
    kind: TNode['kind'];
    leadingTrivia?: LeadingTriviaInput;
    __instructions?: Instructions;
  }
  & {
    -readonly [K in TNodeKeys]: MapTSTypes<Exclude<TNode[K], undefined>>;
  }
  & {
    -readonly [K in Exclude<TPartialNodeKeys, TNodeKeys>]?: MapTSTypes<
      Exclude<TNode[K], undefined>
    >;
  }
>;

export type PickDefinitionFields<T> = Omit<
  T,
  '__instructions' | 'leadingTrivia' | 'kind'
>;

export type DefinitionFields<T> = keyof PickDefinitionFields<T>;
