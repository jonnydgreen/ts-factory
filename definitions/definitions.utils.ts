import type { ts } from '../deps.ts';
import type { Instructions } from '../instructions/instructions.type.ts';
import type { ExpressionInput, MemberNameInput, StatementInput } from './definitions.ts';
import type { LeadingTriviaInput } from './trivia/trivia.type.ts';

export type MergeIntersections<I> = { [K in keyof I]: I[K] };

export type NonFnKeys<T> = {
  // deno-lint-ignore no-explicit-any
  [K in keyof T]: T[K] extends (...args: any[]) => any ? never : K;
}[keyof T];

// deno-fmt-ignore
export type MapTSTypes<T> =
  T extends ts.SyntaxKind ? T :
  T extends ts.NodeArray<infer TNode> ? MapTSTypes<TNode>[] :
  T extends ts.__String ? string :
  T extends string ? string :
  T extends ts.Statement ? StatementInput :
  T extends string | ts.MemberName ? string | MemberNameInput :
  // T extends ts.TypeNode ? TypeNodeInput :
  T extends ts.Expression ? ExpressionInput :
  T extends ts.SyntaxKind ? ExpressionInput :
  unknown;

export type CreateInput<
  TNode extends ts.Node,
  TNodeKeys extends NonFnKeys<TNode>,
  TPartialNodeKeys extends NonFnKeys<TNode> = TNodeKeys,
> = MergeIntersections<
  & {
    kind: TNode['kind'];
    leadingTrivia?: LeadingTriviaInput;
    __instructions?: Instructions;
  }
  & {
    -readonly [K in TNodeKeys]: MapTSTypes<Exclude<TNode[K], undefined>>;
  }
  & {
    -readonly [K in Exclude<TPartialNodeKeys, TNodeKeys>]?: MapTSTypes<
      Exclude<TNode[K], undefined>
    >;
  }
>;
