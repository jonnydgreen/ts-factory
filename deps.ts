// Path
import { join, resolve } from 'https://deno.land/std@0.147.0/path/mod.ts';
export const Path = {
  join,
  resolve,
};

// TS
export { ts } from 'https://deno.land/x/ts_morph@15.1.0/mod.ts';
export * as TSM from 'https://deno.land/x/ts_morph@15.1.0/mod.ts';
