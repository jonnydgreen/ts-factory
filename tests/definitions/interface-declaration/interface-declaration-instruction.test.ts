import { Definition } from '../../../definitions/definitions.ts';
import { ts, tsm } from '../../../deps.ts';
import {
  generateInstructions,
  processInstructions,
} from '../../../instructions/instructions.ts';
import { assertSnapshot, blocks } from '../../../test.deps.ts';
import { createTestName, TestDefinition } from '../../test-utils.ts';

blocks.describe('Interface Declaration', () => {
  const definitions: TestDefinition<Definition>[] = [
    {
      name: createTestName('should process an empty Interface Declaration'),
      input: {
        kind: ts.SyntaxKind.SourceFile,
        statements: [
          {
            kind: ts.SyntaxKind.InterfaceDeclaration,
            name: 'Hello',
            members: [],
          },
        ],
      },
    },
    {
      name: createTestName(
        'should process an Interface Declaration with an Identifier as the name',
      ),
      input: {
        kind: ts.SyntaxKind.SourceFile,
        statements: [
          {
            kind: ts.SyntaxKind.InterfaceDeclaration,
            name: { kind: ts.SyntaxKind.Identifier, text: 'Hello' },
            members: [],
          },
        ],
      },
    },
    {
      name: createTestName('should process an Interface Declaration with modifiers'),
      input: {
        kind: ts.SyntaxKind.SourceFile,
        statements: [
          {
            kind: ts.SyntaxKind.InterfaceDeclaration,
            name: 'Hello',
            modifiers: [{ kind: ts.SyntaxKind.ExportKeyword }],
            members: [],
          },
        ],
      },
    },
    {
      name: createTestName('should process an Interface Declaration with members'),
      input: {
        kind: ts.SyntaxKind.SourceFile,
        statements: [
          {
            kind: ts.SyntaxKind.InterfaceDeclaration,
            name: 'Hello',
            modifiers: [{ kind: ts.SyntaxKind.ExportKeyword }],
            members: [
              {
                kind: ts.SyntaxKind.PropertySignature,
                name: 'foo',
                type: { kind: ts.SyntaxKind.StringKeyword },
              },
              {
                kind: ts.SyntaxKind.PropertySignature,
                name: 'bar',
                type: { kind: ts.SyntaxKind.StringKeyword },
                questionToken: { kind: ts.SyntaxKind.QuestionToken },
              },
            ],
          },
        ],
      },
    },
  ];

  for (const definition of definitions) {
    blocks.it({
      name: definition.name,
      fn: async (t) => {
        // Arrange
        const project = new tsm.Project();
        const sourceFile = project.createSourceFile(
          `${crypto.randomUUID()}.ts`,
          definition.sourceFileContents,
        );
        sourceFile.formatText();
        const instructions = generateInstructions(sourceFile, definition.input);

        // Act
        processInstructions(sourceFile, instructions);

        // Assert
        await assertSnapshot(t, sourceFile.getFullText());
      },
      ignore: definition.ignore,
      only: definition.only,
    });
  }
});
