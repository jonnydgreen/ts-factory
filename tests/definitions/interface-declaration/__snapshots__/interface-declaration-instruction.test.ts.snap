export const snapshot = {};

snapshot[`Interface Declaration > should process an empty Interface Declaration 1`] = `
"interface Hello {
}

"
`;

snapshot[`Interface Declaration > should process an Interface Declaration with an Identifier as the name 1`] = `
"interface Hello {
}

"
`;

snapshot[`Interface Declaration > should process an Interface Declaration with modifiers 1`] = `
"export interface Hello {
}

"
`;

snapshot[`Interface Declaration > should process an Interface Declaration with members 1`] = `
"export interface Hello {
    foo: string;
    bar?: string;
}

"
`;
